src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs

```diff
@@ -24,22 +24,34 @@ public class ArgumentCompleterAttribute : Attribute
     {
         /// <summary/>
         [SuppressMessage("Microsoft.Naming", "CA1721:PropertyNamesShouldNotMatchGetMethods")]
-        public Type Type { get; private set; }
+        public Type Type { get; }
 
         /// <summary/>
-        public ScriptBlock ScriptBlock { get; private set; }
+        public ScriptBlock ScriptBlock { get; }
 
         /// <param name="type">The type must implement <see cref="IArgumentCompleter"/> and have a default constructor.</param>
         public ArgumentCompleterAttribute(Type type)
         {
             if (type == null || (type.GetInterfaces().All(t => t != typeof(IArgumentCompleter))))
             {
-                throw PSTraceSource.NewArgumentException("type");
+                throw PSTraceSource.NewArgumentException(nameof(type));
             }
 
             Type = type;
         }
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ArgumentCompleterAttribute"/> class.
+        /// This constructor is used by derived attributes implementing <see cref="IArgumentCompleterFactory"/>.
+        /// </summary>
+        protected ArgumentCompleterAttribute()
+        {
+            if (GetType().GetInterfaces().All(t => t != typeof(IArgumentCompleterFactory)))
+            {
+                throw PSTraceSource.NewInvalidOperationException();
+            }
+        }
```

Comment:
I still do not understand where the constructor is involved.

Reply:
It is protected to ensure that the only way to call it is from a derived class, and then it fails if the derived class does not implement IArgumentCompleterFactory.

The ctor is implicitly called when the derived class is constructed.

Reply:
Thanks! In the case please add a comment and a test for the scenario.

Reply:
NumberCompletionAttribute ctor calls it implicitly